Cython/Compiler/Parsing.py

```diff
@@ -3967,6 +3974,429 @@ def p_cpp_class_attribute(s, ctx):
             node.decorators = decorators
         return node
 
+def p_match_statement(s, ctx):
+    assert s.sy == "IDENT" and s.systring == "match"
+    pos = s.position()
+    with tentatively_scan(s) as errors:
+        s.next()
+        subject = p_namedexpr_test(s)
+        subjects = None
+        if s.sy == ",":
+            subjects = [subject]
+        while s.sy == ",":
+            s.next()
+            if s.sy == ":":
+                break
+            subjects.append(p_test(s))
+        if subjects is not None:
+            subject = ExprNodes.TupleNode(pos, args=subjects)
+        s.expect(":")
+    if errors:
+        return None
+    # at this stage were commited to it being a match block so continue
+    # outside "with tentatively_scan"
+    # (I think this deviates from the PEG parser slightly, and it'd
+    # backtrack on the whole thing)
+    s.expect_newline()
+    s.expect_indent()
+    cases = []
+    while s.sy != "DEDENT":
+        cases.append(p_case_block(s, ctx))
+    s.expect_dedent()
+    return MatchCaseNodes.MatchNode(pos, subject = subject, cases = cases)
+
+def p_case_block(s, ctx):
+    if not (s.sy=="IDENT" and s.systring == "case"):
+        s.error("Expected 'case'")
+    s.next()
+    pos = s.position()
+    pattern = p_patterns(s)
+    guard = None
+    if s.sy == 'if':
+        s.next()
+        guard = p_test(s)
+    body = p_suite(s, ctx)
+
+    return MatchCaseNodes.MatchCaseNode(pos, pattern=pattern, body=body, guard=guard)
+
+def p_patterns(s):
+    # note - in slight contrast to the name, returns a single pattern
+    patterns = []
+    seq = False
+    pos = s.position()
+    while True:
+        with tentatively_scan(s) as errors:
+            pattern = p_maybe_star_pattern(s)
+        if errors:
+            if patterns:
+                break  # all is good provided we have at least 1 pattern
+            else:
+                e = errors[0]
+                s.error(e.args[1], pos = e.args[0])
+        patterns.append(pattern)
+
+        if s.sy == ",":
+            seq = True
+            s.next()
+            if s.sy in [":", "if"]:
+                break  # common reasons to break
+        else:
+            break
+    if seq:
+        return MatchCaseNodes.MatchSequencePatternNode(pos, patterns = patterns)
+    else:
+        return patterns[0]
+
+def p_maybe_star_pattern(s):
+    # For match case. Either star_pattern or pattern
+    if s.sy == "*":
+        # star pattern
+        s.next()
+        target = None
+        if s.systring != "_":  # for match-case '_' is treated as a special wildcard
+            target = p_pattern_capture_target(s)
+        else:
+            s.next()
+        pattern = MatchCaseNodes.MatchAndAssignPatternNode(
+            s.position(), target = target, is_star = True
+        )
+        return pattern
+    else:
+        p = p_pattern(s)
+        return p
```

Comment:
Looks funny to have two different variables for the same thing, one properly named, one abbreviated.
```suggestion
        pattern = MatchCaseNodes.MatchAndAssignPatternNode(
            s.position(), target = target, is_star = True
        )
        return pattern
    else:
        pattern = p_pattern(s)
        return pattern
```